File: 1-10/5-Longest Palindromic Substring.cpp
//从外往里递归判断是否回文，第92/94案例会超时~~，这个是从中间往外扩展找最大字符串
class Solution {
public:
    bool isLimit(string s, int x){
            if(x>=0 && x<s.size())
                return true;
            return false;
        }
        int extendStr(string s, int start, int end){
            int radius = 0;
            while(isLimit(s,start) && isLimit(s,end)){
                if(s[start]!=s[end])
                    break;
                --start;
                ++end;
                ++radius;
            }
            return radius;
        }
        string longestPalindrome(string s) {
            string ret("");
            int maxlen = 0;
            for(int i=0; i<s.size(); ++i){
                if(s.size()-i +1 <= maxlen/2)
                    break;
                int r1 = extendStr(s,i,i);
                int l1 = r1*2 -1;
                int r2 = extendStr(s,i,i+1);
                int l2 = r2*2;
                if(l1 > l2){
                    if(l1>maxlen){
                        maxlen = l1;
                        ret = s.substr(i-r1+1,l1);
                    }
                }else{
                    if(l2>maxlen){
                        maxlen = l2;
                        ret = s.substr(i-r2+1,l2);
                    }
                }
            }
            return ret;
        }
};

File: py-version/1-10/5. Longest Palindromic Substring.py
class Solution(object):
    def extendStr(self,s,start,end):
        radius = 0
        start -= 1
        end += 1
        while start>-1 and end<len(s) and s[start]==s[end]:
            radius += 1
            start -= 1
            end +=1
        return radius

    def longestPalindrome(self, s):
        ret = ""
        maxlen = 0
        start=0
        end =0
        #print("start")
        for i in range(len(s)):
            if (len(s)-i)<maxlen/2:
                break
            r1 = self.extendStr(s,i,i)
            len1 = r1*2+1
            if len1>maxlen:
                start = i-r1
                end = start + len1
                maxlen = len1
            if i+1<len(s) and s[i]==s[i+1]:
                r2 = self.extendStr(s,i,i+1)
                len2 = r2*2+2
                if len2>maxlen:
                    start = i-r2
                    end = start + len2
                    maxlen = len2
        return s[start:end]
